flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -993,7 +997,7 @@ public MailboxExecutorFactory getMailboxExecutorFactory() {
     }
 
     public boolean canEmitBatchOfRecords() {
-        return !this.mailboxProcessor.hasMail();
+        return !this.mailboxProcessor.hasMail() && taskIsAvailable();
     }
```

Comment:
Could we have a unit test, that would check if this method returns a correct value? Ideally it would be great to have the assertion not only on the `StreamTask#canEmitBatchOfRecords()` level, but on the `DataOutput#...` level - to check that the value is correctly forwarded to the `SourceOperator` and network input.

Reply:
Sure, I can update all code after we discuss.